aten/src/ATen/BatchingRegistrations.cpp

```diff
@@ -354,6 +354,231 @@ Tensor view_as_complex_batching_rule(const Tensor& self) {
   return self_physical.newLogicalFromPhysical(result);
 }
 
+// Checks that the smallest batch stride is greater than the largest example
+// stride. This is something we can support but we choose not to because it's
+// potentially error prone.
+static void checkBatchDimsAtFrontInLayout(IntArrayRef physical_strides, int64_t num_batch_dims) {
+  auto smallest_batch_stride = std::min_element(
+      physical_strides.begin(), physical_strides.begin() + num_batch_dims);
+  auto largest_example_stride = std::max_element(
+      physical_strides.begin() + num_batch_dims, physical_strides.end());
+  if (largest_example_stride == physical_strides.end()) {
+    // No example dimensions
+    return;
+  }
+  TORCH_CHECK(*smallest_batch_stride >= *largest_example_stride,
+    "vmap: Calling Tensor.as_strided is not supported unless the batch dims being ",
+    "vmapped over are at the front of the tensor (in memory layout). When they are ",
+    "not at the front of the tensor this operation can be error prone so we "
+    "actively discourage it; please file us a bug report and/or try to ",
+    "express the as_strided operation in terms of PyTorch view operations");
+}
+
+// given (sizes, strides, storage_offset) returns the maximum location that
+// can be indexed (or nullopt if such a location doesn't exist, e.g., tensors
+// with zero-size dims).
+static optional<int64_t> maximum_indexable_location(
+    IntArrayRef sizes, IntArrayRef strides, int64_t storage_offset) {
+  auto result = native::storage_size_for(sizes, strides);
+  if (result == 0) {
+    return nullopt;
+  }
+  return result + storage_offset;
+}
+
+// Let x be the "first slice" of physical_tensor.
+// This checks that the range of possible memory locations accessible by
+// x.as_strided(sizes, strides, maybe_storage_offset)
+// are within the bounds of possible memory locations accessible by x.
+static void checkBasicAsStridedValidForSlice(
+    const Tensor& physical_tensor,
+    int64_t num_batch_dims,
+    IntArrayRef sizes,
+    IntArrayRef strides,
+    optional<int64_t> maybe_storage_offset) {
+  auto slice_sizes = physical_tensor.sizes().slice(num_batch_dims);
+  auto slice_strides = physical_tensor.strides().slice(num_batch_dims);
+  auto base_offset = physical_tensor.storage_offset();
+
+  auto storage_offset = maybe_storage_offset.value_or(base_offset);
+
+  auto max_as_strided_loc = maximum_indexable_location(sizes, strides, storage_offset);
+  auto max_slice_loc = maximum_indexable_location(slice_sizes, slice_strides, base_offset);
+
+  if (!max_as_strided_loc.has_value()) {
+    return;
+  }
+  if (max_as_strided_loc.has_value() && !max_slice_loc.has_value()) {
```

Comment:
nit: the check for `max_as_strided_loc` having value is not needed here because of the one above.